src/ray/util/event.cc

```diff
@@ -0,0 +1,106 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/util/event.h"
+
+namespace ray {
+
+EventManager &EventManager::Instance() {
+  static EventManager instance_;
+  return instance_;
+}
+
+bool EventManager::IsEmpty() { return reporter_map_.empty(); }
+
+void EventManager::Publish(rpc::Event &event) {
+  for (const auto &element : reporter_map_) {
+    (element.second)->Report(event);
+  }
+}
+
+void EventManager::AddReporter(std::shared_ptr<LogBasedEventReporter> reporter) {
+  reporter_map_.emplace(reporter->GetReporterKey(), reporter);
+}
+
+void EventManager::ClearReporters() { reporter_map_.clear(); }
+
+thread_local std::unique_ptr<RayEventContext> RayEventContext::context_ = nullptr;
+
+RayEventContext &RayEventContext::Instance() {
+  if (context_ == nullptr) {
+    context_ = std::unique_ptr<RayEventContext>(new RayEventContext());
+  }
+  return *context_;
+}
+
+void RayEventContext::SetEventContext(
+    rpc::Event_SourceType source_type,
+    std::unordered_map<std::string, std::string> custom_field) {
+  source_type_ = source_type;
+  custom_field_ = custom_field;
+}
+
+void RayEventContext::ResetEventContext() {
+  source_type_ = rpc::Event_SourceType::Event_SourceType_COMMON;
+  custom_field_.clear();
+}
+
+void RayEventContext::SetCustomField(std::string key, std::string value) {
+  custom_field_[key] = value;
+}
+
+void RayEventContext::SetCustomField(
+    std::unordered_map<std::string, std::string> custom_field) {
+  custom_field_ = custom_field;
+}
+
+void RayEvent::ReportEvent(std::string severity, std::string label, std::string message) {
+  rpc::Event_Severity severity_ele =
+      rpc::Event_Severity::Event_Severity_Event_Severity_INT_MIN_SENTINEL_DO_NOT_USE_;
+  RAY_CHECK(rpc::Event_Severity_Parse(severity, &severity_ele));
+  RayEvent(severity_ele, label) << message;
+}
+
+RayEvent::~RayEvent() { SendMessage(osstream_.str()); }
+
+void RayEvent::SendMessage(const std::string &message) {
+  RAY_CHECK(rpc::Event_SourceType_IsValid(RayEventContext::Instance().GetSourceType()));
+  RAY_CHECK(rpc::Event_Severity_IsValid(severity_));
+
+  if (EventManager::Instance().IsEmpty()) {
+    return;
```

Comment:
Is this condition met before shutdown?

Reply:
Yes, If the user hasn't added event reporters or init the event framework in their own process, the event framework will not send the message. 
BTW, this code can be understood as an optimization. When there is no destination to send, no event class is constructed.